File: src/Integration/Payment/PlategaProvider.php
<?php
declare(strict_types=1);
namespace App\Integration\Payment;

use App\Billing\BillingError;
use Symfony\Component\HttpFoundation\Request;

/**
 * Platega payment provider.
 *
 * Real API (docs.platega.io, project 766087):
 *   POST /v2/transaction/process  -- create a payment link
 *   Auth: header X-MerchantId + X-Secret
 *   Body: paymentDetails{amount,currency}, description, return, failedUrl,
 *         payload, orderId, metadata{userId,userName,clientIp}
 *   Reply: { transactionId, status, url, expiresIn, rate }
 *
 * Amounts use RUB major units; preserve the fractional kopeks.
 */
final class PlategaProvider extends AbstractProvider
{
    private const DEFAULT_BASE = 'https://app.platega.io';

    public function id(): string { return 'platega'; }
    public function name(): string { return 'Platega'; }

    public function configured(): bool
    {
        $enabled = ($this->config['PLATEGA_ENABLED'] ?? '0') === '1'
            || (($this->config['PAYMENT_DRIVER'] ?? '') === 'platega');
        return $enabled
            && ($this->config['PLATEGA_MERCHANT_ID'] ?? '') !== ''
            && ($this->config['PLATEGA_SECRET'] ?? '') !== '';
    }

    private function base(): string
    {
        $base = rtrim((string)($this->config['PLATEGA_API_BASE'] ?? self::DEFAULT_BASE), '/');
        if ($base === '') throw new BillingError('Platega: не задан API-адрес.');
        return $base;
    }

    /**
     * Signed POST to the Platega API. Auth is carried in headers, not in the
     * body, so this is the only place headers are attached.
     */
    private function post(string $endpoint, array $body): array
    {
        $merchant = (string)($this->config['PLATEGA_MERCHANT_ID'] ?? '');
        $secret = (string)($this->config['PLATEGA_SECRET'] ?? '');
        if ($merchant === '' || $secret === '') throw new BillingError('Platega: не заполнены ключи (merchant_id/secret).');
        $resp = $this->json('POST', $this->base().'/'.$endpoint, [
            'headers' => $this->authHeaders(),
            'json' => $body,
        ]);
        if (isset($resp['errors'])) throw new BillingError('Platega: '.$this->summarizeErrors($resp['errors']));
        return $resp;
    }

    private function get(string $endpoint, array $query = []): array
    {
        $merchant = (string)($this->config['PLATEGA_MERCHANT_ID'] ?? '');
        $secret = (string)($this->config['PLATEGA_SECRET'] ?? '');
        if ($merchant === '' || $secret === '') throw new BillingError('Platega: не заполнены ключи (merchant_id/secret).');
        $url = $this->base().'/'.$endpoint;
        if ($query) $url .= '?'.http_build_query($query);
        $resp = $this->json('GET', $url, ['headers' => $this->authHeaders()]);
        if (isset($resp['errors'])) throw new BillingError('Platega: '.$this->summarizeErrors($resp['errors']));
        return $resp;
    }

    private function authHeaders(): array
    {
        return [
            'X-MerchantId' => (string)($this->config['PLATEGA_MERCHANT_ID'] ?? ''),
            'X-Secret' => (string)($this->config['PLATEGA_SECRET'] ?? ''),
            'Content-Type' => 'application/json',
        ];
    }

    private function summarizeErrors($errors): string
    {
        if (is_array($errors)) {
            foreach ($errors as $v) {
                if (is_array($v)) return implode('; ', array_map('strval', $v));
                if (is_string($v) && $v !== '') return $v;
            }
            return 'Некорректные параметры запроса.';
        }
        return is_string($errors) && $errors !== '' ? $errors : 'Некорректные параметры запроса.';
    }

    private function metadata(array $user): array
    {
        $meta = [
            'userId' => (string)($user['telegram_id'] ?? $user['id'] ?? ''),
            'userName' => (string)($user['username'] ?? $user['email'] ?? ''),
        ];
        return array_filter($meta, fn($v) => $v !== '');
    }

    /** POST /v2/transaction/process → check the created payment URL. */
    private function create(array $orderOrTopup, array $user, string $description, string $return, string $failedUrl): array
    {
        $amountMinor = (int)($orderOrTopup['amount_kopeks'] ?? $orderOrTopup['price_minor'] ?? 0);
        if ($amountMinor < 100) throw new BillingError('Platega: некорректная сумма заказа.');
        // Platega expects `amount` in RUB major units (rubles), not kopeks.
        // Sending kopeks here inflates the charge 100x (100 RUB became 10 000 RUB).
        if ($amountMinor < 100) throw new BillingError('Platega: некорректная сумма заказа.');
        $amountRub = self::decimal($amountMinor);
        $res = $this->post('v2/transaction/process', [
            'paymentDetails' => [
                'amount' => $amountRub,
                'currency' => 'RUB',
            ],
            'description' => $description,
            'return' => $return,
            'failedUrl' => $failedUrl,
            'orderId' => (string)$orderOrTopup['id'],
            'payload' => json_encode([isset($orderOrTopup['amount_kopeks'])?'topup_id':'order_id'=>(string)$orderOrTopup['id']],JSON_THROW_ON_ERROR),
            'metadata' => $this->metadata($user),
        ]);
        $url = (string)($res['url'] ?? '');
        if ($url === '' || !str_starts_with($url, 'https://')) throw new BillingError('Platega не вернул ссылку оплаты.');
        $paymentId=$res['transactionId']??null;
        if (!is_string($paymentId) || $paymentId==='' || strlen($paymentId)>100) throw new BillingError('Platega не вернул идентификатор платежа.');
        return [
            'payment_id' => $paymentId,
            'checkout_url' => $url,
        ];
    }

    public function createTopup(array $topup, array $user): array
    {
        $app = rtrim($this->config['APP_URL'] ?? '', '/');
        return $this->create($topup, $user, 'Пополнение баланса', $app.'/balance', $app.'/balance');
    }

    public function createOrder(array $order, array $user): array
    {
        $app = rtrim($this->config['APP_URL'] ?? '', '/');
        return $this->create(
            $order,
            $user,
            'Подписка: '.($order['plan_name'] ?? 'VPN'),
            $app.'/orders/'.$order['id'],
            $app.'/orders/'.$order['id']
        );
    }

    /**
     * Verify a transaction. Statuses observed from the docs: PENDING /
     * CONFIRMED / FAILED / EXPIRED / CANCELED. We query the transaction by
     * its id; the status endpoint returns the same fields as creation.
     */
    public function verify(string $paymentId): array
    {
        // HTTP failures (including 404 and 429) are not payment outcomes.
        // Keep the local payment pending and retry authoritative verification.
        $res = $this->get('transaction/'.rawurlencode($paymentId), []);
        $status = strtoupper((string)($res['status'] ?? ''));
        // Platega returns paymentDetails.amount in RUB rubles (gross) with a
        // separate `comission` field; the merchant-receivable (net) amount =
        // paymentDetails.amount - comission. ZeleBoba settles against the order
        // amount (net), so report the NET amount back in kopeks. Without this, a
        // customer charge of 199.00 + 9% commission (216.91) was read as 21691
        // kopeks vs the order's 19900 -> settle() rejected "Платёж не
        // соответствует заказу" and the Reconciler re-verified forever.
        $amountMinor = self::minor(self::normalizeAmount((string)($res['paymentDetails']['amount'] ?? 0)));
        $commissionMinor = self::minor(self::normalizeAmount((string)($res['comission'] ?? 0)));
        if ($commissionMinor > $amountMinor) throw new BillingError('Platega: некорректная комиссия.');
        $actualId=$res['id']??$res['transactionId']??null;
        if (!is_string($actualId) || $actualId!==$paymentId) throw new BillingError('Platega: несовпадение идентификатора платежа.');
        $metadata=json_decode((string)($res['payload']??'{}'),true);
        if (!is_array($metadata)) $metadata=[];
        if (!isset($metadata['order_id'],$metadata['topup_id']) && !empty($res['orderId'])) $metadata=['order_id'=>(string)$res['orderId'],'topup_id'=>(string)$res['orderId']];
        return [
            'status' => $status === 'CONFIRMED' ? 'paid' : (in_array($status, ['FAILED', 'EXPIRED', 'CANCELED'], true) ? 'canceled' : 'pending'),
            'amount_kopeks' => $amountMinor - $commissionMinor,
            'currency' => (string)($res['paymentDetails']['currency'] ?? 'RUB'),
            'payment_id' => $actualId,
            'metadata' => $metadata,
        ];
    }

    /**
     * Platega callback (webhook) — POST to /webhooks/platega. Payload carries
     * the transaction state; map it to the internal paid/pending/canceled
     * status used by the dispatcher. Returns null if the payload is not a
     * Platega payment callback.
     */
    public function handleWebhook(Request $request): ?array
    {
        // Both authentication headers are required by the Platega callback contract.
        $merchant = (string)($this->config['PLATEGA_MERCHANT_ID'] ?? '');
        $incoming = (string)($request->headers->get('X-MerchantId') ?? $request->headers->get('X-Merchant-Id') ?? '');
        $secret=(string)($this->config['PLATEGA_SECRET']??'');
        if ($merchant==='' || $secret==='' || !hash_equals($merchant,$incoming) || !hash_equals($secret,(string)$request->headers->get('X-Secret',''))) return null;
        $data = $request->toArray();
        $paymentId = (string)($data['transactionId'] ?? $data['id'] ?? $data['payment_id'] ?? '');
        if ($paymentId === '') return null;
        $status = strtoupper((string)($data['status'] ?? ''));
        $orderId = (string)($data['orderId'] ?? $data['order_id'] ?? '');
        return [
            'payment_id' => $paymentId,
            'status' => $status === 'CONFIRMED' ? 'paid' : (in_array($status, ['FAILED', 'EXPIRED', 'CANCELED'], true) ? 'canceled' : 'pending'),
            'metadata' => ['order_id' => $orderId, 'topup_id' => $orderId],
        ];
    }
}

File: rails_backend/app/services/payments/platega_provider.rb
module Payments
  require "erb"
  require "net/http"
  require "uri"

  class PlategaProvider
    DEFAULT_BASE = "https://app.platega.io"

    def id = "platega"

    def config
      @config ||= Infrastructure::RuntimeConfig.new
    end

    def configured?
      enabled = config.fetch("PLATEGA_ENABLED", "0") == "1" || config.fetch("PAYMENT_DRIVER", "") == "platega"
      enabled && config.fetch("PLATEGA_MERCHANT_ID", "").present? && config.fetch("PLATEGA_SECRET", "").present?
    end

    def create_order(order, user)
      app = config.fetch("APP_URL", "").sub(%r{/*\z}, "")
      create(order, user, "Подписка: #{order.plan_name.presence || "VPN"}", "#{app}/orders/#{order.id}", "#{app}/orders/#{order.id}")
    end

    def create_topup(topup, user)
      app = config.fetch("APP_URL", "").sub(%r{/*\z}, "")
      create(topup, user, "Пополнение баланса", "#{app}/balance", "#{app}/balance")
    end

    def verify(payment_id)
      res = request_json(:get, "transaction/#{ERB::Util.url_encode(payment_id)}")
      status = res.fetch("status", "").to_s.upcase
      amount_minor = Amounts.minor(Amounts.normalize(res.dig("paymentDetails", "amount") || "0"))
      commission_minor = Amounts.minor(Amounts.normalize(res["comission"] || "0"))
      raise Billing::BillingError, "Platega: некорректная комиссия." if commission_minor > amount_minor

      actual_id = res["id"] || res["transactionId"]
      raise Billing::BillingError, "Platega: несовпадение идентификатора платежа." unless actual_id == payment_id

      metadata = parse_json(res["payload"])
      if !metadata.key?("order_id") && !metadata.key?("topup_id") && res["orderId"].present?
        metadata = { "order_id" => res["orderId"].to_s, "topup_id" => res["orderId"].to_s }
      end

      {
        "status" => status == "CONFIRMED" ? "paid" : (%w[FAILED EXPIRED CANCELED].include?(status) ? "canceled" : "pending"),
        "amount_kopeks" => amount_minor - commission_minor,
        "currency" => res.dig("paymentDetails", "currency").presence || "RUB",
        "payment_id" => actual_id,
        "metadata" => metadata
      }
    end

    def handle_webhook(request)
      merchant = config.fetch("PLATEGA_MERCHANT_ID", "")
      secret = config.fetch("PLATEGA_SECRET", "")
      incoming = request.headers["X-MerchantId"].presence || request.headers["X-Merchant-Id"].to_s
      return nil if merchant.blank? || secret.blank?
      return nil unless ActiveSupport::SecurityUtils.secure_compare(merchant, incoming)
      return nil unless ActiveSupport::SecurityUtils.secure_compare(secret, request.headers["X-Secret"].to_s)

      data = JSON.parse(request.raw_post)
      payment_id = (data["transactionId"] || data["id"] || data["payment_id"]).to_s
      return nil if payment_id.blank?

      status = data["status"].to_s.upcase
      order_id = (data["orderId"] || data["order_id"]).to_s
      {
        "payment_id" => payment_id,
        "status" => status == "CONFIRMED" ? "paid" : (%w[FAILED EXPIRED CANCELED].include?(status) ? "canceled" : "pending"),
        "event_id" => "#{payment_id}:#{status}",
        "metadata" => { "order_id" => order_id, "topup_id" => order_id }
      }
    rescue JSON::ParserError
      nil
    end

    private

    def create(entity, user, description, return_url, failed_url)
      amount_minor = entity.respond_to?(:amount_kopeks) ? entity.amount_kopeks.to_i : entity.price_minor.to_i
      raise Billing::BillingError, "Platega: некорректная сумма заказа." if amount_minor < 100

      res = request_json(:post, "v2/transaction/process", {
        paymentDetails: { amount: Amounts.decimal(amount_minor), currency: "RUB" },
        description: description,
        return: return_url,
        failedUrl: failed_url,
        orderId: entity.id,
        payload: JSON.generate(entity.respond_to?(:amount_kopeks) ? { topup_id: entity.id } : { order_id: entity.id }),
        metadata: metadata(user)
      })
      checkout_url = res["url"].to_s
      payment_id = res["transactionId"].to_s
      raise Billing::BillingError, "Platega не вернул ссылку оплаты." unless checkout_url.start_with?("https://")
      raise Billing::BillingError, "Platega не вернул идентификатор платежа." if payment_id.blank? || payment_id.length > 100

      { "payment_id" => payment_id, "checkout_url" => checkout_url }
    end

    def metadata(user)
      {
        userId: (user.telegram_id.presence || user.id).to_s,
        userName: (user.try(:username).presence || user.email).to_s
      }.compact_blank
    end

    def request_json(method, endpoint, body = nil)
      raise Billing::BillingError, "Platega: не заполнены ключи (merchant_id/secret)." unless configured?

      uri = URI("#{config.fetch("PLATEGA_API_BASE", DEFAULT_BASE).sub(%r{/*\z}, "")}/#{endpoint}")
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == "https"
      http.open_timeout = 10
      http.read_timeout = 20
      klass = method == :post ? Net::HTTP::Post : Net::HTTP::Get
      req = klass.new(uri)
      req["X-MerchantId"] = config.fetch("PLATEGA_MERCHANT_ID")
      req["X-Secret"] = config.fetch("PLATEGA_SECRET")
      req["Content-Type"] = "application/json"
      req.body = JSON.generate(body) if body
      res = http.request(req)
      raise Billing::BillingError, "Platega returned HTTP #{res.code}." unless res.code.to_i.between?(200, 299)
      JSON.parse(res.body)
    rescue JSON::ParserError
      raise Billing::BillingError, "Platega: некорректный JSON от провайдера."
    end

    def parse_json(value)
      parsed = JSON.parse(value.to_s.presence || "{}")
      parsed.is_a?(Hash) ? parsed : {}
    rescue JSON::ParserError
      {}
    end
  end
end
